compiler/lib/eval.ml

```diff
@@ -84,20 +84,32 @@ let eval_prim x =
   | Ult, [ Int i; Int j ] -> bool (Int32.(j < 0l) || Int32.(i < j))
   | Extern name, l -> (
       let name = Primitive.resolve name in
+      let wrap =
+        match target with
+        | `JavaScript -> fun i -> i
+        | `Wasm -> Int31.wrap
+      in
       match name, l with
       (* int *)
-      | "%int_add", _ -> int_binop l Int.add
-      | "%int_sub", _ -> int_binop l Int.sub
-      | "%direct_int_mul", _ -> int_binop l Int.mul
+      | "%int_add", _ -> int_binop l wrap Int.add
+      | "%int_sub", _ -> int_binop l wrap Int.sub
+      | "%direct_int_mul", _ -> int_binop l wrap Int.mul
       | "%direct_int_div", [ _; Int 0l ] -> None
-      | "%direct_int_div", _ -> int_binop l Int.div
-      | "%direct_int_mod", _ -> int_binop l Int.rem
-      | "%int_and", _ -> int_binop l Int.logand
-      | "%int_or", _ -> int_binop l Int.logor
-      | "%int_xor", _ -> int_binop l Int.logxor
-      | "%int_lsl", _ -> shift l Int.shift_left
-      | "%int_lsr", _ -> shift l Int.shift_right_logical
-      | "%int_asr", _ -> shift l Int.shift_right
+      | "%direct_int_div", _ -> int_binop l wrap Int.div
+      | "%direct_int_mod", _ -> int_binop l wrap Int.rem
+      | "%int_and", _ -> int_binop l wrap Int.logand
+      | "%int_or", _ -> int_binop l wrap Int.logor
+      | "%int_xor", _ -> int_binop l wrap Int.logxor
+      | "%int_lsl", _ -> shift l wrap Fun.id Int.shift_left
+      | "%int_lsr", _ ->
+          shift
+            l
+            wrap
+            (match target with
+            | `JavaScript -> Fun.id
+            | `Wasm -> fun i -> Int.logand i 0x7fffffffl)
```

Comment:
Why do we need this logand ?

Reply:
I’m not sure. @vouillon, could you explain why?

Reply:
The number `-1` is represented as the 32-bit integer `0xffffffff`, with all bits sets. If we shift it to the right by one, we should get the integer `0x3fffffff` (30 bits remain set), but without clearing the sign bit first we get the incorrect value `0x7fffffff`.

Reply:
Maybe we should introduce a new module Int_target and use it inside constant `Int of Int_target.t

Reply:
> The number `-1` is represented as the 32-bit integer `0xffffffff`, with all bits sets.

It seems to be one of the possible choices. I would be curious to know the rationale for it.

Reply:
If you leave the sign bit cleared, some operations, like `(asr)`, `(/)`, or simply printing an integer, get more complicated.

Reply:
Looking more at this, I wonder if it would be simpler to 
- shift_right by 1
- do the Int32 op
- shift_left by 1

Reply:
Similar to how Int63.t is emulated using Int64.t in https://github.com/janestreet/base/blob/master/src/int63_emul.ml, 
we could have Int31.t implemented as a int32 shilfted to the left.

Reply:
This does not look that much simpler to me. There are still a lot of special cases for `(*)`, `(/)`, `(lsr)`, `(asr)`, ...

Reply:
> , with all bits sets. If we shift it to the right by one, we should get the integer `0x3fffffff` (30 bits remain set), but without clearing the sign bit first we get the incorrect value `0x7fffffff`.

I somehow find reasoning about this (2 complement logic) much harder than the understanding theses special cases.
Let's move the discussion next the new implementation in stdlib.